internal/session/sessionwants.go

```diff
@@ -89,21 +108,39 @@ func (sw *sessionWants) BlocksReceived(ks []cid.Cid) ([]cid.Cid, time.Duration)
 		}
 	}
 
+	// If the live wants ordering array is a long way out of sync with the
+	// live wants map, clean up the ordering array
+	if len(sw.liveWantsOrder)-len(sw.liveWants) > liveWantsOrderGCLimit {
+		cleaned := sw.liveWantsOrder[:0]
+		for _, c := range sw.liveWantsOrder {
+			if _, ok := sw.liveWants[c]; ok {
+				cleaned = append(cleaned, c)
+			}
+		}
+		sw.liveWantsOrder = cleaned
+	}
```

Comment:
Note: a possibly simpler (and probably more efficient) alternative is to just have two sets: targeted and broadcasting.

* When we receive a block, we remove it from both sets.
* When we prepare a broadcast, we move CIDs from the targeted set to the broadcast set as necessary.

Reply:
Note: this _does_ ignore order, but that shouldn't really matter anyways.

Reply:
The reason I wanted to have ordering is because when we get an idle tick we
- broadcast pending wants to connected peers
- query the DHT for the first pending want

For both broadcast and DHT query we will likely get better results if we choose wants closer to the front of the list (because more peers will likely have nodes in higher branches in the DAG)

Reply:
Fair enough.